### ISSUE TITLE
Test failure: test_only_update_old_event

### ISSUE THREAD

**Issue by lfam** [pimutils/khal#309]
Hi,

I'm trying to package khal for GNU Guix [0], but the build process fails during the test phase on ```test_only_update_old_event```.

Specifically, the assertion that ```cal._db_needs_update()``` fails.

I've attached the build log: [khal-0.7.0-build.txt](https://github.com/geier/khal/files/60379/khal-0.7.0-build.txt)

Please let me know if I can provide more information or help in any other way.

Here's the logged output of the test suite, cut from the full log:
```
starting phase `check'
============================= test session starts ==============================
platform linux -- Python 3.4.3 -- py-1.4.23 -- pytest-2.6.1
plugins: cov
collected 154 items

tests/aux_test.py ............
tests/backend_test.py ......................
tests/cal_display_test.py ....x
tests/cli_test.py .....x........
tests/controller_test.py ...
tests/event_test.py .............................
tests/khalendar_aux_test.py ................................
tests/khalendar_test.py .....................F
tests/settings_test.py .....
tests/terminal_test.py ...
tests/vtimezone_test.py ...
tests/ui/test_calendarwidget.py ...
tests/ui/test_widgets.py .

=================================== FAILURES ===================================
__________________________ test_only_update_old_event __________________________

cal_vdir = (<khal.khalendar.khalendar.Calendar object at 0x7ffff1e28518>, <FilesystemStorage(**{'path': '/tmp/nix-build-khal-0.7.0.drv-0/pytest-0/test_only_update_old_event0'})>)
monkeypatch = <_pytest.monkeypatch.monkeypatch object at 0x7ffff0f7b748>

    def test_only_update_old_event(cal_vdir, monkeypatch):
        cal, vdir = cal_vdir
    
        href_one, etag_one = vdir.upload(cal.new_event(dedent("""
        BEGIN:VEVENT
        UID:meeting-one
        DTSTART;VALUE=DATE:20140909
        DTEND;VALUE=DATE:20140910
        SUMMARY:first meeting
        END:VEVENT
        """)))
    
        href_two, etag_two = vdir.upload(cal.new_event(dedent("""
        BEGIN:VEVENT
        UID:meeting-two
        DTSTART;VALUE=DATE:20140910
        DTEND;VALUE=DATE:20140911
        SUMMARY:second meeting
        END:VEVENT
        """)))
    
        cal.db_update()
        assert not cal._db_needs_update()
    
        old_update_vevent = cal._update_vevent
        updated_hrefs = []
    
        def _update_vevent(href):
            updated_hrefs.append(href)
            return old_update_vevent(href)
        monkeypatch.setattr(cal, '_update_vevent', _update_vevent)
    
        href_three, etag_three = vdir.upload(cal.new_event(dedent("""
        BEGIN:VEVENT
        UID:meeting-three
        DTSTART;VALUE=DATE:20140911
        DTEND;VALUE=DATE:20140912
        SUMMARY:third meeting
        END:VEVENT
        """)))
    
>       assert cal._db_needs_update()
E       assert <bound method Calendar._db_needs_update of <khal.khalendar.khalendar.Calendar object at 0x7ffff1e28518>>()
E        +  where <bound method Calendar._db_needs_update of <khal.khalendar.khalendar.Calendar object at 0x7ffff1e28518>> = <khal.khalendar.khalendar.Calendar object at 0x7ffff1e28518>._db_needs_update

tests/khalendar_test.py:410: AssertionError
=============== 1 failed, 151 passed, 2 xfailed in 59.90 seconds ===============
phase `check' failed after 60.6 seconds
```

Here's a sorted list of all the dependencies being used to build khal and its dependencies, excluding implicit dependencies such as GCC, tar, make, etc.
```
acl-2.2.52
attr-2.4.46
bash-4.3.39
bison-2.7.1
bison-3.0.4
bzip2-1.0.6
coreutils-8.24
dbus-1.10.0
expat-2.1.0
file-5.22
flex-2.5.37
fontconfig-2.11.94
freetype-2.6
gdbm-1.11
gettext-0.19.6
glib-2.46.1
gmp-6.0.0a
gs-fonts-8.11
hicolor-icon-theme-0.12
indent-2.2.10
inputproto-2.3.1
intltool-0.51.0
kbproto-1.0.6
khal-0.7.0
libcap-2.24
libffi-3.1
libgcrypt-1.6.3
libgpg-error-1.19
libpthread-stubs-0.3
libx11-1.6.2
libxau-1.0.8
libxcb-1.11
libxdmcp-1.1.1
libxext-1.3.3
libxft-2.3.2
libxml2-2.9.3
libxrender-0.9.8
libxslt-1.1.28
m4-1.4.17
ncurses-6.0
openssl-1.0.2e
perl-5.16.1
perl-xml-parser-2.44
pkg-config-0.28
python-3.4.3
python-apipkg-1.4
python-atomicwrites-0.1.8
python-blinker-1.3
python-cffi-1.2.1
python-click-6.2
python-click-log-0.1.1
python-click-threading-0.1.2
python-configobj-5.0.6
python-coverage-3.7.1
python-cryptography-1.0.2
python-cryptography-vectors-1.0.2
python-dateutil-2.4.2
python-docutils-0.12
python-enum34-1.0
python-execnet-1.4.1
python-httplib2-0.9.1
python-icalendar-3.9.1
python-idna-2.0
python-iso8601-0.1.10
python-jinja2-2.7.3
python-jsonschema-2.4.0
python-lxml-3.4.2
python-markupsafe-0.23
python-minimal-3.4.3
python-mock-1.0.1
python-nose-1.3.4
python-oauthlib-1.0.3
python-pretend-1.0.8
python-psutil-3.0.1
python-py-1.4.23
python-pyasn1-0.1.8
python-pycparser-2.10
python-pygments-1.6
python-pyjwt-1.4.0
python-pytest-2.6.1
python-pytest-cache-1.0
python-pytest-cov-2.2.0
python-pytest-localserver-0.3.4
python-pytest-runner-2.6.2
python-pytest-xprocess-0.9.1
python-pytz-2015.7
python-pyxdg-0.25
python-requests-2.8.1
python-requests-toolbelt-0.5.0
python-setuptools-18.3.1
python-setuptools-scm-1.9.0
python-six-1.10.0
python-sphinx-1.2.3
python-sphinxcontrib-newsfeed-0.1.4
python-tzlocal-1.2
python-urwid-1.2.2
python-vdirsyncer-0.7.3
python-werkzeug-0.11.2
python-wheel-0.26.0
python-wrapper-3.4.3
python-wrapper-3.4.3
python-wsgi-intercept-0.10.3
radicale-1.0.1
readline-6.3
renderproto-0.11.1
shared-mime-info-1.2
sqlite-3.8.11.1
tcl-8.6.4
tk-8.6.4
tzdata-2015c
unzip-6.0
util-macros-1.19.0
xcb-proto-1.11
xextproto-7.3.0
xproto-7.0.26
xtrans-1.3.5
zlib-1.2.7
```

[0]
http://www.gnu.org/software/guix/

**Comment by geier** [comment]
Thank you @lfam for building a package and reporting this bug, I'll have a closer look ASAP.

**Comment by geier** [comment]
Sorry for not getting back to you earlier. I have finally been able to reproduce this (sometimes), but I haven't really found the cause yet and therefore not fixed it yet. But I haven't forgotten about this.

**Comment by lfam** [comment]
Interesting! It seems to be completely reproducible within the Guix build system.

**Comment by lfam** [comment]
By the way, I added the package:
https://www.gnu.org/software/guix/package-list.html#khal

If you click on the package's name you can see the packaging, which disables this test.

Here is an image of Khal's dependency graph, excluding implicit dependencies like GCC and the tar used to unpack the source tarball:
http://imgur.com/xgsq0j4

**Comment by lfam** [comment]
Reading the new bugs I noticed https://github.com/geier/khal/issues/320. I wonder if this bug has a similar cause.

**Comment by geier** [comment]
I believe you are right and those are related.  Could you perhaps either test the current MASTER or patch v.0.7.0 so that lines 395 in tests/khalendar_test.py contains `sleep(1)`?

**Comment by lfam** [comment]
The build fails when setuptools-scm can't figure out the version of something (Khal itself?). Any idea how to work around that?

```starting phase `build'
running "python setup.py" with command "build" and parameters ()
Traceback (most recent call last):
  File "setup.py", line 54, in <module>
    "Topic :: Communications",
  File "/gnu/store/13n8xbi9wv9pigfyhir007qadr81jq46-python-3.4.3/lib/python3.4/distutils/core.py", line 108, in setup
    _setup_distribution = dist = klass(attrs)
  File "/gnu/store/13n8xbi9wv9pigfyhir007qadr81jq46-python-3.4.3/lib/python3.4/site-packages/setuptools/dist.py", line 269, in __init__
    _Distribution.__init__(self,attrs)
  File "/gnu/store/13n8xbi9wv9pigfyhir007qadr81jq46-python-3.4.3/lib/python3.4/distutils/dist.py", line 280, in __init__
    self.finalize_options()
  File "/gnu/store/13n8xbi9wv9pigfyhir007qadr81jq46-python-3.4.3/lib/python3.4/site-packages/setuptools/dist.py", line 325, in finalize_options
    ep.load()(self, ep.name, value)
  File "/gnu/store/x7zdfc3b3hc50npbgnisy4ydmbiwgl76-python-setuptools-scm-1.9.0/lib/python3.4/site-packages/setuptools_scm-1.9.0-py3.4.egg/setuptools_scm/integration.py", line 19, in version_keyword
  File "/gnu/store/x7zdfc3b3hc50npbgnisy4ydmbiwgl76-python-setuptools-scm-1.9.0/lib/python3.4/site-packages/setuptools_scm-1.9.0-py3.4.egg/setuptools_scm/__init__.py", line 77, in get_version
  File "/gnu/store/x7zdfc3b3hc50npbgnisy4ydmbiwgl76-python-setuptools-scm-1.9.0/lib/python3.4/site-packages/setuptools_scm-1.9.0-py3.4.egg/setuptools_scm/__init__.py", line 36, in version_from_scm
LookupError: setuptools-scm was unable to detect version for '/tmp/nix-build-khal-0.7.0.drv-0/source'.

Make sure you're not using GitHub's tarballs (or similar ones), as those don't contain the necessary metadata. Use PyPI's tarballs instead.

For example, if you're using pip, instead of https://github.com/user/proj/archive/master.zip use git+https://github.com/user/proj.git#egg=proj
phase `build' failed after 0.7 seconds
```

**Comment by untitaker** [comment]
>Make sure you're not using GitHub's tarballs (or similar ones), as those don't contain the necessary metadata. Use PyPI's tarballs instead.

**Comment by lfam** [comment]
On Mon, Jan 18, 2016 at 03:41:58PM -0800, Markus Unterwaditzer wrote:
> >Make sure you're not using GitHub's tarballs (or similar ones), as those don't contain the necessary metadata. Use PyPI's tarballs instead.

I am not using a GitHub tarball. I'm building from a checkout of HEAD.
Are there extra steps I need to take to make such a checkout buildable?

**Comment by untitaker** [comment]
You need to have git installed, and the checkout must still be a git repo (i.e.
./.git must still be there)

**Comment by lfam** [comment]
Ah, that is impossible with Guix, because the .git subdirectory is non-deterministic and thus ruins the content-addressable-ness of the Nix / Guix system.

I will try geier's other suggestion, patching the relevant file directly. I wanted to build from HEAD because seems that a lot has changed since the last release.

**Comment by lfam** [comment]
I used Guix to check out the sources of 0.7.0. This involves deleting the ```.git``` metadata as mentioned above. I initialized a new git repository in the unpacked sources and generated the patch below with `git-format-patch`.

The test suite passes with this change. There are still two "xfails" that don't seem to cause the whole thing to fail. I didn't report them earlier because I figured they weren't important if the build process succeeds without them. They are:
```
tests/cal_display_test.py ....x
tests/cli_test.py .....x........
```
Patch:
```
From 707d9638df05b4b97fa8cedb72dd35b654fc4aa9 Mon Sep 17 00:00:00 2001
From: lfam
Date: Mon, 18 Jan 2016 18:52:50 -0500
Subject: [PATCH] Sleep for debugging test failure

---
 tests/khalendar_test.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/tests/khalendar_test.py b/tests/khalendar_test.py
index 5040df6..e5e2ab1 100644
--- a/tests/khalendar_test.py
+++ b/tests/khalendar_test.py
@@ -393,6 +393,8 @@ def test_only_update_old_event(cal_vdir, monkeypatch):
     old_update_vevent = cal._update_vevent
     updated_hrefs = []
 
+    sleep(1)
+
     def _update_vevent(href):
         updated_hrefs.append(href)
         return old_update_vevent(href)
-- 
2.6.4
```

**Comment by geier** [comment]
To get the last cal_display_test running, you need to install the `cs_CZ.UTF-8` locale, the xfailing cli_test is currently xfailing everywhere because it tries to set the current date to a predetermined date and this doesn't work yet.

I will soon release a new version, either 0.7.1 with only this change or 0.8.0, but I'm not sure yet.

**Comment by lfam** [comment]
While reading ```tests/cal_display_test```, I noticed that the relevant test's name is ```test_vertical_month_unicode_weekdeays```. Is ```weekdeays``` a typo?

Also, is it testing something specific to the Czech locale, or can I patch it to use some other locale?

I ask because the full suite of locales is a very heavy dependency (450 MB) for users using Guix on distros besides GuixSD. Guix has a second locale package (```glibc-utf8-locales```) (6 MB) that is often used for testing; it contains ```"de_DE" "el_GR" "en_US" "fr_FR" "tr_TR"```.

What if I patched the test to use something from that list?

**Comment by geier** [comment]
We are using Czech because their weekname abbreviations contain non-ascii characters. Turkish or Greek might work as well, I'll have a look tonight.

And yes, that is a typo.

**Comment by hobarrera** [comment]
> Ah, that is impossible with Guix, because the .git subdirectory is non-deterministic and thus ruins the content-addressable-ness of the Nix / Guix system.

I believe that a shallow clone of a certain tag should always be identical, maybe you can use that?

**Comment by lfam** [comment]
On Tue, Jan 19, 2016 at 05:18:26AM -0800, Hugo Osvaldo Barrera wrote:
> > Ah, that is impossible with Guix, because the .git subdirectory is non-deterministic and thus ruins the content-addressable-ness of the Nix / Guix system.
> 
> I believe that a shallow clone of a certain tag should always be identical, maybe you can use that?

Perhaps. I believe this was discussed on the guix-devel mailing list
recently and some workarounds were suggested. But it is not a big
limitation in this situation — patching the file from the release
tarball was sufficient.

**Comment by geier** [comment]
I've just added a test for the greek locale, it really shouldn't matter which one you run.

**Comment by lfam** [comment]
On Tue, Jan 19, 2016 at 04:09:06PM -0800, Christian Geier wrote:
> I've just added a test for the greek locale, it really shouldn't matter which one you run.

Thanks!

Do you want me to submit the "sleep 1" patch from upthread or do you
want to deal with the issues with this test some other way?

**Comment by geier** [comment]
At least with current master, sleep(1) shouldn't be necessary, something like sleep(0.01) should suffice. I'll need to experiment on that. So, you don't need to submit the patch, and I really hope that in the next release you won't have to patch the code at all.

**Comment by lfam** [comment]
On Wed, Jan 20, 2016 at 02:41:05AM -0800, Christian Geier wrote:
> At least with current master, sleep(1) shouldn't be necessary, something like sleep(0.01) should suffice. I'll need to experiment on that. So, you don't need to submit the patch, and I really hope that in the next release you won't have to patch the code at all.

Thanks! I'll wait for the next release.

**Comment by untitaker** [comment]
Have we agreed on using `sleep` in the tests? I have another instance where tests are failing locally.

**Comment by geier** [comment]
Do you have a better idea?

**Comment by untitaker** [comment]
No I don't think so

On 3 February 2016 14:24:22 CET, Christian Geier <notifications@github.com> wrote:
>Do you have a better idea?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/geier/khal/issues/309#issuecomment-179226898

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by untitaker** [comment]
I pushed a commit to master that fixes another testfailure (at least I had one
locally).

**Comment by geier** [comment]
Thanks, I believe that one also errored several times on travis.

Quoting Markus Unterwaditzer (2016-02-04 20:35:13)

**Comment by geier** [comment]
@lfam this one, too?

**Comment by lfam** [comment]
Yes, this one is also fixed in 0.8.1!

**Comment by geier** [comment]
thank you for confirming!

Quoting lfam (2016-04-26 18:36:11)
